zypp/PurgeKernels.cc

```diff
@@ -0,0 +1,422 @@
+
+      std::string kernelVersion;
+      std::ifstream procKernel( "/proc/sys/kernel/osrelease" );
+      if ( procKernel ) {
+        std::getline( procKernel, kernelVersion );
+      }
+      return kernelVersion;
+
+    }
+
+
+    std::set<size_t>  _keepLatestOffsets = { 0 };
+    std::set<size_t>  _keepOldestOffsets;
+    std::set<Edition> _keepSpecificEditions;
+    std::string       _uname_r;
+    Arch              _kernelArch;
+    std::string       _keepSpec = ZConfig::instance().multiversionKernels();
+    bool              _keepRunning    = true;
+  };
+
+  /*!
+   * tries to remove a the \ref PoolItem \a pi from the pool, solves and checks if no unexpected packages are removed due to the \a validRemovals regex.
+   * If the constraint fails the changes are reverted and \a false is returned.
+   */
+  bool PurgeKernels::Impl::removePackageAndCheck( PoolItem &pi, const str::regex &validRemovals ) const
+  {
+    const filter::ByStatus toBeUninstalledFilter( &ResStatus::isToBeUninstalled );
+    auto pool = ResPool::instance();
+
+    //remember which packages are already marked for removal, we do not need to check them again
+    std::set< sat::Solvable::IdType> currentSetOfRemovals;
+    for ( auto it = pool.byStatusBegin( toBeUninstalledFilter ); it != pool.byStatusEnd( toBeUninstalledFilter );  it++  )
+      currentSetOfRemovals.insert( it->id() );
+
+    pi.status().setToBeUninstalled( ResStatus::USER );
+
+    if ( !pool.resolver().resolvePool() ) {
+      MIL << "Failed to resolve pool, skipping " << pi << std::endl;
+      pool.resolver().problems();
+      pi.statusReset();
+
+      return false;
+    }
+
+    for ( auto it = pool.byStatusBegin( toBeUninstalledFilter ); it != pool.byStatusEnd( toBeUninstalledFilter );  it++  ) {
+
+      //this was set by us or marked by a previous removal, ignore them
+      if ( it->status().isByUser() || (currentSetOfRemovals.find( it->id() ) != currentSetOfRemovals.end()) )
+        continue;
+
+      str::smatch what;
+      if ( !str::regex_match( it->name(), what, validRemovals) ) {
+        MIL << "Package " << PoolItem(*it) << " should not be removed, skipping " << pi << std::endl;
+        pi.statusReset();
+        return false;
+      }
+    }
+
+    MIL << "Removing package: " << pi << std::endl;
+    return true;
+  }
+
+  /*!
+   * Parse the config line keep spec that tells us which kernels should be kept
+   */
+  void PurgeKernels::Impl::parseKeepSpec( )
+  {
+    //keep spec parse regex, make sure to edit the group offsets if changing this regex
+    const str::regex specRegex( "^(latest|oldest)([+-][0-9]+)?$", str::regex::match_extended );
+
+    const unsigned tokenGrp = 1; //index of the group matching the token
+    const unsigned modifierGrp = 2; //index of the group matching the offset modifier
+
+
+    MIL << "Parsing keep spec: " << _keepSpec << std::endl;
+
+    std::vector<std::string> words;
+    str::split( _keepSpec, std::back_inserter(words), "," );
```

Comment:
I'd relax the parser to trim the words. Actually I'd enhance str::split to accept an optional `Trim` argument:
```
    template<class TOutputIterator>
      unsigned split( const C_Str & line_r, TOutputIterator result_r, const C_Str & sepchars_r = " \t", const Trim trim_r = NO_TRIM )
    template<class TOutputIterator>
      unsigned split( const C_Str & line_r, TOutputIterator result_r, const Trim trim_r )
      { return split( line_r, result_r, " \t", trim_r ); }
```
May cherry-pick b4343234febc303c6f655e6b2905c4346cb49b11

Reply:
Done